### PR TITLE
[Data][Docs] Split "Consuming data" guide

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

The "Consuming data" guide is long and contains several unrelated topics. For better discoverability, this PR splits "Consuming data" into three separate guides:
* "Inspecting data"
* "Iterating over data"
* "Saving data"

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/data/iterating-over-data.rst [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+:ref:`Transforming batches <transforming-batches>` and
+:ref:`Inspecting batches <inspecting-batches>`.
+
+.. _iterating-over-batches-with-shuffling:
+
+Iterating over batches with shuffling
+=====================================
+
+:class:`Dataset.random_shuffle <ray.data.Dataset.random_shuffle>` is slow. For better
+performance, shuffle data while iterating.
```
Comment (amogkam) [inline doc/source/data/iterating-over-data.rst:142]: can you add more information here? I think we want to explicitly say that this is a local shuffle, bounded by the provided buffer size.

Also why is `random_shuffle` slow? Should we also link to this section from the `random_shuffle` API reference?

File: doc/source/data/iterating-over-data.rst [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+To iterate over batches with shuffling, specify ``local_shuffle_buffer_size``.
+
+.. tip::
+
+    To configure ``local_shuffle_buffer_size``, choose the largest value that fits
+    in memory. For most workloads, 10x the batch size is reasonable.
```
Comment (amogkam) [inline doc/source/data/iterating-over-data.rst:149]: Is the 10x always true?

File: doc/source/data/saving-data.rst [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+=============================
+
+To save your :class:`~ray.data.dataset.Dataset` to local filesystems, call a method
+like :meth:`~ray.data.write_csv` and specify a local directory with the `local://`
+scheme.
+
+.. warning::
+
+    If your cluster contains multiple nodes and you don't use `local://`, Ray Data
+    writes your data to multiple nodes.
```
Comment (amogkam) [inline doc/source/data/saving-data.rst:17]: Clarify this some more- write different partitions of data to different nodes.

File: doc/source/data/saving-data.rst [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    import ray
+
+    ds = ray.data.read_csv("example://iris.csv")
+
+    ds.write_csv("local:///tmp/iris/")
+
+To write data to formats other than CSV, read the :ref:`Input/Output reference <input-output>`.
+
+Writing data to S3
```
Comment (amogkam) [inline doc/source/data/saving-data.rst:29]: s3 -> cloud storage? google cloud storage is also supported

File: doc/source/data/saving-data.rst [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+To save your :class:`~ray.data.dataset.Dataset` to S3, call a method like
+:meth:`~ray.data.write_csv` and specify a bucket or folder.
+
+.. testcode::
+
+    import ray
+
+    ds = ray.data.read_csv("s3://anonymous@air-example-data/iris.csv")
+
+    ds.write_csv("s3://my-bucket/my-folder")
```
Comment (amogkam) [inline doc/source/data/saving-data.rst:41]: Add a tab for gcs?

File: doc/source/data/saving-data.rst [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    ds = ray.data.read_csv("example://iris.csv")
+
+    ds.write_csv("/mnt/cluster_storage/iris")
+
+To write data to formats other than CSV, read the :ref:`Input/Output reference <input-output>`.
+
+Changing the number of output files
+===================================
+
+When you call a write method, Ray Data writes your data to one file per *block*. To
```
Comment (amogkam) [inline doc/source/data/saving-data.rst:65]: are we still planning to have block in key concept? should we link to that?

**Review by amogkam — COMMENTED** [review]
overall looks good, left some comments

File: doc/source/data/inspecting-data.rst [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+          sepal width (cm): double,
+          petal length (cm): double,
+          petal width (cm): double,
+          target: int64
+       }
+    )
+
+.. _inspecting-rows:
+
+Inspecting rows
```
Comment (amogkam) [inline doc/source/data/inspecting-data.rst:59]: also mention take_all somewhere?

**Review by amogkam — COMMENTED** [review]
We're missing the following consumption patterns:
1. Converting to a different in-memory data format (`to_pandas`, `to_spark`, `to_modin`, etc.). 
2. Splitting and consuming the dataset

File: doc/source/data/inspecting-data.rst [hunk 8]
```diff
@@ -0,0 +1,130 @@
+.. _inspecting-data:
+
+===============
+Inspecting data
+===============
+
```
Comment (amogkam) [inline doc/source/data/inspecting-data.rst:6]: can you add a sentence or 2 about what this guide covers? Just consistent pattern with the other user guides.

File: doc/source/data/inspecting-data.rst [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+          target: int64
+       }
+    )
+
+.. _inspecting-rows:
+
+Inspecting rows
+===============
+
+To inspect rows, call :meth:`Dataset.take() <ray.data.Dataset.take>` or
```
Comment (amogkam) [inline doc/source/data/inspecting-data.rst:62]: Just add a little bit more on what this is doing? To inspect rows, call ..., which returns a list of rows from the Dataset.

File: doc/source/data/inspecting-data.rst [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+:meth:`Dataset.take_all() <ray.data.Dataset.take_all>`. Ray Data represents rows
```
Comment (amogkam) [inline doc/source/data/inspecting-data.rst:63]:
```suggestion
:meth:`Dataset.take_all() <ray.data.Dataset.take_all>`. Ray Data represents each row as a dictionary
```

File: doc/source/data/inspecting-data.rst [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+:ref:`Iterating over rows <iterating-over-rows>`.
+
+.. _inspecting-batches:
+
+Inspecting batches
+==================
+
+A batch contains data from multiple rows. To inspect batches, call
+`Dataset.take_batch() <ray.data.Dataset.take_batch>`. By default, Ray Data represents
+batches as ndarrays.
```
Comment (amogkam) [inline doc/source/data/inspecting-data.rst:91]: Dictionary of Numpy ndarray. You can set the `"batch_format"` to change the format of the returned batch.

File: doc/source/data/inspecting-data.rst [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+            import ray
+
+            ds = ray.data.read_images("example://image-datasets/simple")
+
+            batch = ds.take_batch(batch_size=2, batch_format="numpy")
+            print(batch)
+
+        .. testoutput::
+
+            {'image': array([[[[...]]]], dtype=uint8)}
```
Comment (amogkam) [inline doc/source/data/inspecting-data.rst:108]: can we show something more meaningful as the output here?

possibly using a different dataset if images are too complicated

File: doc/source/data/iterating-over-data.rst [hunk 13]
```diff
@@ -0,0 +1,273 @@
+.. _iterating-over-data:
+
+===================
+Iterating over data
+===================
+
```
Comment (amogkam) [inline doc/source/data/iterating-over-data.rst:6]: Add an intro sentence or 2 here to match with the other user guides?

File: doc/source/data/iterating-over-data.rst [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+.. _iterating-over-rows:
+
+Iterating over rows
+===================
+
+To iterate over rows, call :meth:`Dataset.iter_rows() <ray.data.Dataset.iter_rows>`. Ray
```
Comment (amogkam) [inline doc/source/data/iterating-over-data.rst:12]: To iterate over the rows of your dataset,

File: doc/source/data/iterating-over-data.rst [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+Data represents rows as dictionaries.
```
Comment (amogkam) [inline doc/source/data/iterating-over-data.rst:13]: represents each row as a dictionary

File: doc/source/data/iterating-over-data.rst [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+For more information on working with rows, see
+:ref:`Transforming rows <transforming-rows>` and
+:ref:`Inspecting rows <inspecting-rows>`.
+
+.. _iterating-over-batches:
+
+Iterating over batches
+======================
+
+A batch contains data from multiple rows. To iterate over batches, call one of the
```
Comment (amogkam) [inline doc/source/data/iterating-over-data.rst:41]: Iterate over batches of dataset in different formats by calling one of the following methods:

File: doc/source/data/iterating-over-data.rst [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+
+            tf_dataset = ds.to_tf(
+                feature_columns="sepal length (cm)",
+                label_columns="target",
+                batch_size=2
+            )
+            for features, labels in tf_dataset:
+                print(features, labels)
+
+        .. testoutput:
```
Comment (amogkam) [inline doc/source/data/iterating-over-data.rst:126]:
```suggestion
        .. testoutput::
```

File: doc/source/data/iterating-over-data.rst [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+:ref:`Transforming batches <transforming-batches>` and
+:ref:`Inspecting batches <inspecting-batches>`.
+
+.. _iterating-over-batches-with-shuffling:
+
+Iterating over batches with shuffling
+=====================================
+
+:class:`Dataset.random_shuffle <ray.data.Dataset.random_shuffle>` is slow because it
+shuffles all rows. For better performance, shuffle a subset of rows while iterating.
```
Comment (amogkam) [inline doc/source/data/iterating-over-data.rst:142]: when would I want to use `random_shuffle` then?

I think we should clarify this more "If a full global shuffle is not required, you can shuffle a subset of rows up to a provided buffer size during iteration by specifying `local_shuffle_buffer_size`. While this is not a true global shuffle like `random_shuffle`, it is more performant as it does not require excessive data movement"

File: doc/source/data/iterating-over-data.rst [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+====================================================
+
+If you're performing distributed data parallel training, call
+:meth:`Dataset.split <ray.data.Dataset.split>` to split your dataset into disjoint
+shards.
+
+.. note::
+
+  If you're using :ref:`Ray Train <train-docs>`, you don't need to split the dataset.
+  Ray Train automatically splits your dataset for you.
```
Comment (amogkam) [inline doc/source/data/iterating-over-data.rst:257]: link to the AIR configurig ingest guide.

File: doc/source/data/saving-data.rst [hunk 20]
```diff
@@ -0,0 +1,198 @@
+.. _saving-data:
+
+===========
+Saving data
+===========
+
```
Comment (amogkam) [inline doc/source/data/saving-data.rst:6]: add intro sentence for consistency

File: doc/source/data/iterating-over-data.rst [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+
+            tf.Tensor([6.1 6.3], shape=(2,), dtype=float64) tf.Tensor([1 2], shape=(2,), dtype=int64)
+            ...
+            tf.Tensor([6.1 6.3], shape=(2,), dtype=float64) tf.Tensor([1 2], shape=(2,), dtype=int64)
+
+Splitting datasets for distributed parallel training
+====================================================
+
+If you're performing distributed data parallel training, call
+:meth:`Dataset.split <ray.data.Dataset.split>` to split your dataset into disjoint
```
Comment (ericl) [inline doc/source/data/iterating-over-data.rst:251]: streaming_split

File: doc/source/data/saving-data.rst [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+    If your cluster contains multiple nodes and you don't use `local://`, Ray Data
+    writes different partitions of data to different nodes.
+
+.. testcode::
+
+    import ray
+
+    ds = ray.data.read_csv("example://iris.csv")
+
+    ds.write_parquet("local:///tmp/iris/")
```
Comment (amogkam) [inline doc/source/data/saving-data.rst:28]: does this code snippet still get tested even if there's not testoutput::?

File: doc/source/data/saving-data.rst [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+To write data to formats other than Parquet, read the
+:ref:`Input/Output reference <input-output>`.
+
+Writing data to cloud storage
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To save your :class:`~ray.data.dataset.Dataset` to cloud storage, authenticate all nodes
+with your cloud service provider. Then, call a method like
+:meth:`Dataset.write_parquet <ray.data.Dataset.write_parquet>` and specify a bucket or
+folder URI.
```
Comment (amogkam) [inline doc/source/data/saving-data.rst:39]: what other cloud providers are supported?

File: doc/source/data/saving-data.rst [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+:ref:`Input/Output reference <input-output>`.
+
+Converting Datasets to other Python libraries
+=============================================
+
+Converting Datasets to pandas
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To convert a :class:`~ray.data.dataset.Dataset` to a pandas DataFrame, call
+:meth:`Dataset.to_pandas() <ray.data.Dataset.to_pandas>`.
```
Comment (amogkam) [inline doc/source/data/saving-data.rst:93]: Add a note that this requires your dataset to fit in memory on a single node.

**Review by amogkam — APPROVED** [review]
looks good! just some minor comments

File: doc/source/data/saving-data.rst (hunk 22 above)
Comment (bveeramani) [reply doc/source/data/saving-data.rst:28]: Yes

File: doc/source/data/saving-data.rst (hunk 23 above)
Comment (bveeramani) [reply doc/source/data/saving-data.rst:39]: Just Azure, right? I've added a tab for Azure

**Comment by amogkam** [comment]
@bveeramani I think you forgot to push the changes?

**Pull request merged**